### PR TITLE
Extract Ack-related logic to a separate module

### PR DESCRIPTION
**Pull request by jciupis**

### CODE REVIEW

File: src/nrf_802154_ack_pending_bit.h [hunk 1]
```diff
@@ -30,6 +30,8 @@
 
 /**
  * @brief This file implements procedures to set pending bit in nRF 802.15.4 radio driver.
```
Comment (KaSroka) [inline src/nrf_802154_ack_pending_bit.h]: Add `@deprecated`

File: src/nrf_802154_core.c [hunk 2]
```diff
@@ -1918,8 +1901,15 @@ static void irq_crcok_state_rx(void)
             bool wait_for_phyend;
 
             // Prepare ACK
-            ack_prepare();
-            nrf_radio_packet_ptr_set(m_ack_psdu);
+            mp_ack = nrf_802154_ack_generator_create(mp_current_rx_buffer->psdu);
+
+            // TODO: What to do in case of NULL?
```
Comment (KaSroka) [inline src/nrf_802154_core.c]: I think this have to be sorted out before merge.

File: src/mac_features/ack/nrf_802154_ack_generator.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    nrf_802154_imm_ack_generator_init();
+    nrf_802154_enh_ack_generator_init();
+}
+
+const uint8_t * nrf_802154_ack_generator_create(const uint8_t * p_frame)
+{
+    // This function should not be called if ACK is not requested.
+    assert(p_frame[ACK_REQUEST_OFFSET] & ACK_REQUEST_BIT);
+
+    (void)frame_version_is_2015_or_above(p_frame);
```
[the remark this answers is not in the document]
Comment (jciupis) [reply src/mac_features/ack/nrf_802154_ack_generator.c]: It's deprecated now, I'll remove this call.

File: src/mac_features/nrf_802154_frame_parser.c [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+/**
+ * @file
+ *   This file implements frame parsing utilities for 802.15.4 radio driver.
+ *
+ * @note This module is based on following assumptions:
+ *       a. All received frames contain both source and destination address.
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_frame_parser.c]: We cannot assume it. We can receive an invalid frame. We must accept any type of frame. If receive frame mode is not supported, we should just drop it.
Perhaps this module is designed correctly, but I would change this description to make clear that frame is just dropped if any of assumptions is not met.

File: src/mac_features/nrf_802154_frame_parser.c [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+const uint8_t * nrf_802154_frame_parser_dst_addr_get(const uint8_t * p_frame, bool * p_dst_addr_extended)
+{
+    if (!addressing_is_valid(p_frame))
+    {
+        *p_dst_addr_extended = false;
+        return NULL;
+    }
+    else
+    {
+        *p_dst_addr_extended = nrf_802154_frame_parser_dst_addr_is_extended(p_frame);
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_frame_parser.c]: What if dst addr mode is None

File: src/mac_features/nrf_802154_frame_parser.c [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+const uint8_t * nrf_802154_frame_parser_src_addr_get(const uint8_t * p_frame, bool * p_src_addr_extended)
+{
+    if (!addressing_is_valid(p_frame))
+    {
+        *p_src_addr_extended = false;
+        return NULL;
+    }
+    else
+    {
+        *p_src_addr_extended = nrf_802154_frame_parser_src_addr_is_extended(p_frame);
```
Comment (hubertmis) [inline src/mac_features/nrf_802154_frame_parser.c]: Looks that None mode is not handle here as well.

File: src/nrf_802154.h [hunk 7]
```diff
@@ -911,6 +913,22 @@ void nrf_802154_pan_coord_set(bool enabled);
...
+ * @param[in]  extended If the given address is an extended MAC address or a short MAC address.
+ * @param[in]  p_data   Pointer to the buffer containing data to be set.
+ * @param[in]  length   Length of @p p_data.
+ * 
+ * @retval True   Address successfully added to the list.
+ * @retval False  There is not enough memory to store this address in the list.
+ */
+bool nrf_802154_ack_data_set(const uint8_t * p_addr,
+                             bool            extended,
+                             const uint8_t * p_data,
```
Comment (hubertmis) [inline src/nrf_802154.h]: I think this param should be `const void *` to make it generic.

File: src/nrf_802154.h [hunk 8]
```diff
@@ -911,6 +913,22 @@ void nrf_802154_pan_coord_set(bool enabled);
  */
 bool nrf_802154_pan_coord_get(void);
 
+/**
+ * @brief Add address of a peer node for which provided ACK data should be set.
+ *
+ * @param[in]  p_addr   Array of bytes containing the address of the node (little-endian).
+ * @param[in]  extended If the given address is an extended MAC address or a short MAC address.
+ * @param[in]  p_data   Pointer to the buffer containing data to be set.
```
Comment (hubertmis) [inline src/nrf_802154.h]: There should be one more parameter that describes what kind of data is updated (like `uint32_t data_type` enumerating pending bit, IE, etc). Also this parameter may be included in the p_data structure, but then this structure should be defined.

File: src/nrf_802154_const.h [hunk 9]
```diff
@@ -88,14 +112,24 @@
 #define SRC_ADDR_OFFSET_SHORT_DST    8         ///< Offset of source address in Data frame if destination address is short
 #define SRC_ADDR_OFFSET_EXTENDED_DST 14        ///< Offset of source address in Data frame if destination address is extended
 
-#define PHR_SIZE              1    ///< Size of PHR field
+#define ACK_LENGTH            5    ///< Length of ACK frame
```
Comment (hubertmis) [inline src/nrf_802154_const.h]: Imm-Ack frame

File: src/nrf_802154_core.c (hunk 2 above)
Comment (hubertmis) [inline src/nrf_802154_core.c]: Just go to `else` in line 2015. Perhaps `if` above should be modified to make it clear.

File: src/mac_features/ack/nrf_802154_enh_ack_generator.c [hunk 10]
```diff
@@ -0,0 +1,393 @@
...
+
+#include <assert.h>
+#include <string.h>
+
+#include "mac_features/nrf_802154_frame_parser.h"
+#include "nrf_802154_ack_pending_bit.h"
+#include "nrf_802154_const.h"
+#include "nrf_802154_enh_ack_generator.h"
+
+#define ENH_ACK_MAX_SIZE     MAX_PACKET_SIZE
```
[the remark this answers is not in the document]
Comment (jciupis) [reply src/mac_features/ack/nrf_802154_enh_ack_generator.c]: How does `MAX_IE_SIZE` affect Enh-Ack buffer size?

[the remark this answers is not in the document]
Comment (hubertmis) [reply src/mac_features/ack/nrf_802154_enh_ack_generator.c]: Enh-Ack size is sth like: FCF + seq num + extended addresses + 2 pan id + security header + mic-128 + _MAX_IE_SIZE_ + FSC

**Comment by hubertmis** [comment]
Closing this one, replacing with new one with resolved conflicts, fixed bugs and added features.

**Pull request closed without merging**